### PR TITLE
Cassandra 17658

### PR DESCRIPTION
**Pull request by jacek-lewandowski**

### CODE REVIEW

File: src/java/org/apache/cassandra/db/Keyspace.java [hunk 1]
```diff
@@ -137,7 +140,10 @@ public static void setInitialized()
     @VisibleForTesting
     public static void unsetInitialized()
     {
-        initialized = false;
+        synchronized (Schema.instance)
```
Comment (adelapena) [inline src/java/org/apache/cassandra/db/Keyspace.java:143]: I think the method `Kesypace.unsetInitialized` isn't used anywhere, and it didn't seem to be used when it was introduced for the first time on CASSANDRA-17044. Should we remove it?

File: src/java/org/apache/cassandra/schema/Schema.java [hunk 2]
```diff
@@ -227,6 +227,7 @@ public Keyspace maybeAddKeyspaceInstance(String keyspaceName, Supplier<Keyspace>
         return keyspaceInstances.blockingLoadIfAbsent(keyspaceName, loadFunction);
     }
 
+    @VisibleForTesting
```
Comment (adelapena) [inline src/java/org/apache/cassandra/schema/Schema.java]: The method `Schema.maybeRemoveKeyspaceInstance` doesn't seem used anymore, since it has lost its single caller, `Schema.dropKeyspace`. Should we remove it?

File: test/long/org/apache/cassandra/cql3/CorruptionTest.java [hunk 3]
```diff
@@ -59,7 +60,11 @@ public class CorruptionTest extends SchemaLoader
     @BeforeClass()
     public static void setup() throws ConfigurationException, IOException
     {
-        Schema.instance.clear();
```
Comment (adelapena) [inline test/long/org/apache/cassandra/cql3/CorruptionTest.java:62]: Nit: removing this leaves an unused import of `Schema`.

File: test/long/org/apache/cassandra/cql3/CorruptionTest.java [hunk 4]
```diff
@@ -59,7 +60,11 @@ public class CorruptionTest extends SchemaLoader
...
+        DatabaseDescriptor.daemonInitialization();
+
+        // clean-up is only needed if we run the test from IDE
+        ServerTestUtils.mkdirs();
+        ServerTestUtils.cleanup();
```
Comment (adelapena) [inline test/long/org/apache/cassandra/cql3/CorruptionTest.java]: This is always done before any instantiation of `EmbeddedCassandraService`, so we may make tests slightly briefer and less error prone if we encapsulate both things into a single utility method. Going a bit further, we could move the entire `EmbeddedCassandraService` class into the tests directory (it's only used for  tests), and place the cleanup of dirs into its constructor. wdyt?

Comment (jacek-lewandowski) [reply test/long/org/apache/cassandra/cql3/CorruptionTest.java]: I was thinking about that, I agree we can safely have a single method which includes `mkdirs` and `cleanup`., or even have a static methods which does that and creates ECS. However, I'm not sure we should move `EmbeddedCassandraService` to the tests directory because it is currently a user visible thing and some users may start Cassandra that way. When it was moved into the tests directory, it would not be a part of the `cassandra-all` artifact in public Maven repository. This is also why we should not put `mkdirs` and `cleanup` into the constructor

Comment (adelapena) [reply test/long/org/apache/cassandra/cql3/CorruptionTest.java]: Right, there could be someone using `EmbeddedCassandraService` from the Maven artifact. Then we can put the repeated block of code into a test method into a utility method, more or less [this way](https://github.com/adelapena/cassandra/commit/c6f25e16f548c929bc79e6e91d9efc8df6ad3124). I think that would save us some code duplication, wdyt?

File: src/java/org/apache/cassandra/schema/Schema.java (hunk 2 above)
Comment (adelapena) [reply src/java/org/apache/cassandra/schema/Schema.java]: After last changes the only caller of `Schema.maybeRemoveKeyspaceInstance` is 
 `Schema.dropKeyspace`, so we shouldn't add the `@VisibleForTesting` annotation. We could also make the method private.

Comment (ekaterinadimitrova2) [reply src/java/org/apache/cassandra/schema/Schema.java]: I think this comment might have been missed

File: test/long/org/apache/cassandra/cql3/CorruptionTest.java (hunk 4 above)
Comment (ekaterinadimitrova2) [reply test/long/org/apache/cassandra/cql3/CorruptionTest.java]: I agree with everything said here, maybe I would just rename the method as we already have start in `EmbeddedCassandraService` itself

File: src/java/org/apache/cassandra/schema/Schema.java [hunk 5]
```diff
@@ -533,13 +535,13 @@ private synchronized void updateVersion(UUID version)
     }
 
     /**
-     * Clear all KS/CF metadata and reset version.
+     * Clear all KS/CF metadata and reset version. Not to be ever used in production code
      */
-    public synchronized void clear()
+    @VisibleForTesting
+    public synchronized void clearUnsafe()
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/schema/Schema.java]: My IDE didn't find any usages of this method. Also, searching on the commit I didn't find any?

File: src/java/org/apache/cassandra/schema/Schema.java [hunk 6]
```diff
@@ -626,7 +628,9 @@ public void resetLocalSchema()
         updateHandler.clear();
 
         logger.debug("Clearing local schema keyspace instances...");
-        clear();
+        distributedKeyspaces.forEach(this::unload);
+        updateVersion(SchemaConstants.emptyVersion);
+        SchemaDiagnostics.schemaCleared(this);
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/schema/Schema.java:622]: Those were in a synchronized method before?

File: test/unit/org/apache/cassandra/ServerTestUtils.java [hunk 7]
```diff
@@ -189,6 +190,16 @@ public static void cleanupSavedCaches()
         cleanupDirectory(DatabaseDescriptor.getSavedCachesLocation());
     }
 
+    public static EmbeddedCassandraService startEmbeddedCassandraService() throws IOException
```
Comment (ekaterinadimitrova2) [inline test/unit/org/apache/cassandra/ServerTestUtils.java:193]: I am wondering whether this is not a bit of a confusing name. We already have a start method in EmbeddedCassandraService.

File: test/unit/org/apache/cassandra/schema/SchemaTest.java [hunk 8]
```diff
@@ -43,19 +44,16 @@
     public static void setupDatabaseDescriptor()
     {
         DatabaseDescriptor.daemonInitialization();
+        ServerTestUtils.prepareServer();
+        Schema.instance.loadFromDisk();
```
Comment (ekaterinadimitrova2) [inline test/unit/org/apache/cassandra/schema/SchemaTest.java:46]: `loadFromDisk `is not exactly a DD setup, maybe it should be just a `setup `method and not `setupDatabaseDescriptor`?

File: test/unit/org/apache/cassandra/schema/SchemaTest.java [hunk 9]
```diff
@@ -43,19 +44,16 @@
...
     }
 
     @Test
     public void testTransKsMigration() throws IOException
     {
-        CommitLog.instance.start();
```
Comment (ekaterinadimitrova2) [inline test/unit/org/apache/cassandra/schema/SchemaTest.java:51]: Now we can clean the unused import import `org.apache.cassandra.db.commitlog.CommitLog;`

**Review by ekaterinadimitrova2 — COMMENTED** [review]
Overall LGTM, left just a few small comments/questions

File: src/java/org/apache/cassandra/schema/Schema.java (hunk 6 above)
Comment (jacek-lewandowski) [reply src/java/org/apache/cassandra/schema/Schema.java:622]: good point, I need to revisit this

File: test/unit/org/apache/cassandra/ServerTestUtils.java (hunk 7 above)
Comment (jacek-lewandowski) [reply test/unit/org/apache/cassandra/ServerTestUtils.java:193]: Why confusing? `start` is instance method in `EmbeddedCassandraService`. This is a different class, dedicated solely for setting up Cassandra instance for testing purposes

File: src/java/org/apache/cassandra/schema/Schema.java (hunk 2 above)
Comment (jacek-lewandowski) [reply src/java/org/apache/cassandra/schema/Schema.java]: We discussed on Slack, but I can apply this comment

Comment (ekaterinadimitrova2) [reply src/java/org/apache/cassandra/schema/Schema.java]: I am not sure what was discussed tbh, I think it was with @adelapena probably

File: test/unit/org/apache/cassandra/ServerTestUtils.java (hunk 7 above)
Comment (ekaterinadimitrova2) [reply test/unit/org/apache/cassandra/ServerTestUtils.java:193]: Because it doesn't do only the start from what I see, but at the same time I guess cleaning should be expected part of the start process in the tests. I am not super worried, just expressed opinion :-) I guess we can just add a Java doc, I will leave it to you

File: src/java/org/apache/cassandra/schema/Schema.java (hunk 6 above)
Comment (ekaterinadimitrova2) [reply src/java/org/apache/cassandra/schema/Schema.java:622]: Hmm, do we want the whole method now? I mean before it was considered that only those three operations were enough. Just trying to be careful with the price of synchronization but at the same time not to miss it :-)

File: src/java/org/apache/cassandra/schema/DefaultSchemaUpdateHandler.java [hunk 10]
```diff
@@ -257,12 +255,11 @@ private synchronized SchemaTransformationResult reload()
...
-               ? reload()
-               : migrationCoordinator.pullSchemaFromAnyNode()
-                                     .flatMap(mutations -> ImmediateFuture.success(applyMutations(mutations)))
-                                     .awaitThrowUncheckedOnInterrupt()
-                                     .getNow();
+        if (local)
+            return reload();
+
+        Collection<Mutation> mutations = migrationCoordinator.pullSchemaFromAnyNode().awaitThrowUncheckedOnInterrupt().getNow();
+        return applyMutations(mutations);
```
Comment (ekaterinadimitrova2) [inline src/java/org/apache/cassandra/schema/DefaultSchemaUpdateHandler.java:262]: Not sure why this change was needed

File: src/java/org/apache/cassandra/schema/Schema.java (hunk 6 above)
Comment (jacek-lewandowski) [reply src/java/org/apache/cassandra/schema/Schema.java:622]: Yes, I think it was a mistake that it wasn't synchronized. Synchronization does not really matter in terms of performance in this case because that method is called only from NodeTool and it is one time action. Basically what we need I think is that there are no other schema changes issued locally when we are restoring schema from the other node, not just during cleaning

File: src/java/org/apache/cassandra/schema/DefaultSchemaUpdateHandler.java (hunk 10 above)
Comment (jacek-lewandowski) [reply src/java/org/apache/cassandra/schema/DefaultSchemaUpdateHandler.java:262]: Yeah, so in the previous scenario, `reset` was called from `Schema.resetLocalSchema`, in a thread 1. In this method, we pull the schema from some node, which is actually processed asynchronously in thread 2 and returns `Future`. Hence we use `flatMap`, `applyMutations` was called in thread 2. It ended up in a deadlock because `Schema.resetLocalSchema`, which is now synchronized was waiting for the `reset` to finish, while `applyMutations` called from thread 2 was waiting for updating `Schema` in thread 2. 

Since it is blocking anyway, we firstly fetch the schema using whichever thread, and then continue with `applyMutations` in the main thread which initiated the schema changes

Comment (ekaterinadimitrova2) [reply src/java/org/apache/cassandra/schema/DefaultSchemaUpdateHandler.java:262]: Oh, yeah, needed because now the whole method was synchronized, side effect... thanks :-)
